keepassxc_browser/cli.py

```diff
@@ -0,0 +1,177 @@
+#!/usr/bin/python3
+
+import sys
+import os
+import re
+import subprocess
+import argparse
+import json
+import tldextract
+from . import protocol as kpp
+
+
+def find_pass_candidates(domain, connection, auth_id):
+    try:
+        candidates = connection.get_logins(auth_id,"https://"+domain)
+    except kpp.ProtocolError as e:
+        print("failure: KeePassXC protocol error:", e, file=sys.stderr)
+        sys.exit(1)
+
+    return {
+        "%s [%s]" % (entry["name"], entry["login"]): (entry["name"], entry["login"], entry["password"]) for entry in candidates
+    }
+
+
+def select_candidate(items, command):
+    process = subprocess.run(command, input='\n'.join(items).encode("utf-8"),
+                             stdout=subprocess.PIPE, shell=True)
+    return process.stdout.decode("utf-8").strip()
+
+
+def fetch_candidates(url, auth_id, keyfile):
+    conn = kpp.Connection()
+    try:
+        conn.connect()
+    except Exception as e:
+        print("failure: cannot connect to KeePassXC server:", e, file=sys.stderr)
+        sys.exit(1)
+
+    keyfile_name = os.path.expanduser(keyfile)
+
+    try:
+        if os.path.exists(keyfile_name):
+            try:
+                keyfile = open(keyfile_name, "r")
+                cred = kpp.Identity.unserialize(auth_id, keyfile.read())
+                keyfile.close()
+            except Exception as e:
+                print("failure: error when try to read keyfile", e, file=sys.stderr)
+                sys.exit(1)
+
+            conn.change_public_keys(cred)
+
+            if not conn.test_associate(cred):
+                conn.associate(cred)
+        else:
+            cred = kpp.Identity(auth_id)
+            conn.change_public_keys(cred)
+            conn.associate(cred)
+            try:
+                keyfile = open(keyfile_name, "w")
+                keyfile.write(cred.serialize())
+                keyfile.close()
+            except Exception as e:
+                print("failure: error when trying to write keyfile", e, file=sys.stderr)
+                sys.exit(1)
+    except kpp.ProtocolError as e:
+        print("failure: KeePassXC protocol error:", e, file=sys.stderr)
+        sys.exit(1)
+
+    extract_result = tldextract.extract(url)
+
+    # Try to find candidates using targets in the following order: fully-qualified domain name (includes subdomains),
```

Comment:
What's the thought process here? You try to do some sensible matching on the url? What usecase are you covering? At least this needs to be documented (see below @argparse) somehow.

Reply:
That is in parts from the qutebrowser scripts for old KeePass. The idea is, if no match is found for the given domain, to search 'upward' the tree of subdomains, up to the last before the TLD.